### PR TITLE
add role assignment check for `azd pipeline config`

### LINKED ISSUE #1954: Command "azd pipeline config" failing due to unauthorized error, even though I'm a contributor to the Azure subscription I'm using
- [x] Make sure you've installed the latest version using [instructions in the wiki](../wiki/install)

**Output from `azd version`**
Run `azd version` and copy and paste the output here:
azd version 0.8.0-beta.1 (commit 08209da7c22d28c6a123ef4e88d433132066b800)

**Describe the bug**

I create an empty folder and used the `azd init --template AzureSamples/todo-python-mongo-aca` command to initialized my sample application. `azd up` worked as expected and I got the app deployed. However, when running `azd pipeline config`, it errored out with the following message:
```
(base) PS REDACTED\Apps\pycon-azd-todo> azd pipeline config

Configure your azd pipeline

  (✓) Done: Checking current directory for Git repository
  |       |          Creating or updating service principal <REDACTED>
  (x) Failed: Creating or updating service principal <REDACTED>

ERROR: failed to create or update service principal: failed applying role assignment: failed assigning role assignment 'REDACTED' to service principal 'az-dev-04-14-2023-22-56-35' : PUT https://management.azure.com/subscriptions/REDACTED/providers/Microsoft.Authorization/roleAssignments/REDACTED
--------------------------------------------------------------------------------  
RESPONSE 403: 403 Forbidden
ERROR CODE: AuthorizationFailed
--------------------------------------------------------------------------------  
{
  "error": {
    "code": "AuthorizationFailed",
    "message": "The client 'REDACTED' with object id 'REDACTED' does not have authorization or an ABAC condition not fulfilled to perform action 'Microsoft.Authorization/roleAssignments/write' over scope '/subscriptions/1REFACTED/providers/Microsoft.Authorization/roleAssignments/REDACTED' or the scope is invalid. If access was recently granted, please refresh your credentials."
  }
}
--------------------------------------------------------------------------------  
```
I double checked and I am a contributor to the subscription I used.
 
**To Reproduce**
I ran the following in an empty folder:
> `azd auth login`
>  `azd init --template Azure-Samples/todo-python-mongo-aca` 
> `azd up`
> `azd pipeline config`

**Environment**
Information on your environment:
    * Language name and version: Python 3.9
    * IDE and version : VS Code 1.78.0-insider

### PR DESCRIPTION
**Pull request by hemarina**
fix #1954

### CODE REVIEW

File: cli/azd/pkg/output/ux/list_as_text.go [hunk 1]
```diff
@@ -14,7 +14,7 @@ import (
 // ["foo", "bar"] 					=> "foo and bar"
 // ["foo", "bar", "axe"] 			=> "foo, bar and axe"
 // ["foo", "bar", ..., ..., "axe"] 	=> "foo, bar, ..., ... and axe"
-func ListAsText(items []string) string {
+func ListAsText(items []string, connectWord string) string {
```
Comment (vhvb1989) [inline cli/azd/pkg/output/ux/list_as_text.go:17]: Can we have `AndListAsText()` and `OrListAsText()` ?   Both should call the same function where the `and` or `or` is passed and then add test cases for both.
Keep the implementation function private like `listAsText([]string, string)`.

File: cli/azd/pkg/tools/azcli/ad.go [hunk 2]
```diff
@@ -55,8 +56,10 @@ func (cli *azCli) CreateOrUpdateServicePrincipal(
 		return nil, fmt.Errorf("failed resetting application credentials: %w", err)
 	}
 
+	// Required role assignment for applying role assignment
+	checkRoles := []string{"Owner", "User Access Administrator"}
 	// Apply specified role assignment
-	err = cli.ensureRoleAssignments(ctx, subscriptionId, roleName, servicePrincipal)
+	err = cli.ensureRoleAssignments(ctx, subscriptionId, roleName, servicePrincipal, checkRoles)
```
Comment (vhvb1989) [inline cli/azd/pkg/tools/azcli/ad.go:62]: `ensureRoleAssignments` is too late to check if the user has the required roles.
This method is about adding the roles to the service principal after it is created.   We don't want to create the service principal if we know we can't assign the roles.

The roles validation must be performed before calling `CreateOrUpdateServicePrincipal` , as early as possible and before doing any other operation

File: cli/azd/pkg/output/ux/list_as_text.go [hunk 3]
```diff
@@ -14,7 +14,7 @@ import (
 // ["foo", "bar"] 					=> "foo and bar"
 // ["foo", "bar", "axe"] 			=> "foo, bar and axe"
 // ["foo", "bar", ..., ..., "axe"] 	=> "foo, bar, ..., ... and axe"
-func ListAsText(items []string) string {
+func listAsText(items []string, connectWord string) string {
```
Comment (vhvb1989) [inline cli/azd/pkg/output/ux/list_as_text.go:17]: mind updating the comments above for the method?  You might want to move the docs to `AndListAsText` .
Then add docs for `OrListAsText`.

Optionally, add some docs for `listAsText`  (since this is not public)

File: cli/azd/pkg/tools/azcli/ad.go [hunk 4]
```diff
@@ -350,3 +356,20 @@ func (cli *azCli) createRoleAssignmentsClient(
 
 	return client, nil
 }
+
+func (cli *azCli) CheckRoleAssignments(
```
Comment (vhvb1989) [inline cli/azd/pkg/tools/azcli/ad.go:360]: This is not OK.

`cli.getRoleDefinition` checks if a definition exists for some scope.  For example, we use if to check if the role `Owner` exists for a `Resouce Group` or for a `Subscription`.  This method will not tell who this role is assigned to. It will ONLY help to ensure a role can be assigned to someone.

For example, if I want to assign the role `Foo` to a person `Bar`, for the resource `XYZ`, I can first get the role `Foo` from `XYZ`.  And if the role is found, then I can try to assign it to `Bar`.  But if the role is not found within `XYZ`, then there would be an error trying to assign it to someone.

We need `CheckRoleAssignments()` to take a Principal-id and use it to get the role assignments for that principal-id and the resource.  For the previous example, the function would take `Bar` id and `XYZ`.  It will then return all the roles that are assigned to `Bar` for `XYZ`.  Then someone can loop the list and check if `Foo` is part of the list or not

File: cli/azd/pkg/commands/pipeline/pipeline_manager.go [hunk 5]
```diff
@@ -114,6 +114,14 @@ func (i *PipelineManager) preConfigureCheck(ctx context.Context, infraOptions pr
 		)
 	}
 
+	// Required authorized role assignments
+	checkRoles := []string{"Owner", "User Access Administrator"}
+	// Check permission on role assignment
+	err = i.azCli.CheckRoleAssignments(ctx, i.Environment.GetSubscriptionId(), checkRoles)
+	if err != nil {
+		return configurationWasUpdated, err
+	}
```
Comment (wbreza) [inline cli/azd/pkg/commands/pipeline/pipeline_manager.go:123]: In my opinion we should NOT be enforcing that the service principal has `Owner` or `User Access Administrator` for `azd pipeline config` to be run.  

The required roles solely depends on the underlying apps requirements.  Yes - anything that performs role assignments in the IaC will require these but many apps do not.

I would instead invest the time in providing a better error message when we see these types of ARM deployment failures and instruct the user what to do.  This is easily correct by the user by providing the `--principal-role owner` argument to the `azd pipeline config` command.

Consider adding similar troubleshooting tips in the template readme if it doesn't exist already.
https://github.com/azure-samples/todo-nodejs-mongo-aks#troubleshootingknown-issues

**Review by wbreza — CHANGES_REQUESTED** [review]

**Comment by wbreza** [comment]
@hemarina - this PR may no longer be required after #2218 since we are updating the default role to `Contributor` + `User Access Administrator` that will resolve the RBAC role assignment issues in some templates.

**Comment by vhvb1989** [comment]
> @hemarina - this PR may no longer be required after #2218 since we are updating the default role to `Contributor` + `User Access Administrator` that will resolve the RBAC role assignment issues in some templates.

We still need to check that whoever runs `azd pipeline config` now has either owner or User Admin to fail fast.  If a contributor-only runs `azd pipeline config`, the operation will fail in the middle (after creating the gh repo) when trying to assign the roles to the SP

File: cli/azd/pkg/tools/azcli/ad.go [hunk 6]
```diff
@@ -367,3 +373,105 @@ func (cli *azCli) createRoleAssignmentsClient(
...
+		if err != nil {
+			return nil, fmt.Errorf("failing to get role definition name from role definition id: %w", err)
+		}
+		roleDefinitionName = append(roleDefinitionName, *roleDefinitionResult.Properties.RoleName)
+	}
+
+	return roleDefinitionName, nil
+}
+
+func (cli *azCli) CheckRoleAssignments(
```
Comment (ellismg) [inline cli/azd/pkg/tools/azcli/ad.go:442]: It feels to me that this is the wrong place for this method to live.  As a matter of policy, `pipeline configure` wants to ensure that before it starts the current principal has either the "Owner" or "User Access Administrator" policies applied to it, because if it does not, it wants to warn the user that the operation may fail later when it tries to assign roles to the newly created service principal.  That feels like logic that is very tied to the policy of the pipeline config gesture, and so it should live in the pipeline manager.

I would export getUserRoleDefinitionName and keep it on this type, so that the code here in CheckRoleAssignments that will move into `pipeline_manager.go` can call it.

**Review by ellismg — CHANGES_REQUESTED** [review]
Over all this is looking very good.  As things evolved, I think we ended up with some code living in a place it shouldn't, so there's a comment about moving some stuff away from the azcli type and into pipeline_manager where I think the policy belongs.

File: cli/azd/pkg/tools/azcli/ad.go [hunk 7]
```diff
@@ -367,3 +372,68 @@ func (cli *azCli) createRoleAssignmentsClient(
 
 	return client, nil
 }
+
+// Get role definition id from role assignment within specific scope and principalID
+func (cli *azCli) getUserRoleDefinitionID(
```
Comment (wbreza) [inline cli/azd/pkg/tools/azcli/ad.go:377]: Since this method returns a list of role ids I would consider changing the function name to something like 'getUserRoleDefinitionIdList` or similar.

File: cli/azd/pkg/commands/pipeline/pipeline_manager.go [hunk 8]
```diff
@@ -420,3 +446,42 @@ func (manager *PipelineManager) Configure(ctx context.Context) (
...
+func checkRoleAssignments(
+	ctx context.Context,
+	azcli azcli.AzCli,
+	subscriptionId string,
+	principalId string,
+	console input.Console,
+) error {
+	// Find the specified role in the subscription scope
+	scope := azure.SubscriptionRID(subscriptionId)
+	roles, err := azcli.GetUserRoleDefinitionName(ctx, subscriptionId, scope, principalId)
```
Comment (wbreza) [inline cli/azd/pkg/commands/pipeline/pipeline_manager.go:459]: With this current implementation there are going to be other edge case conditions that we are not accounting for.  Users could have the correct permissions but may be inherited via custom role definition or roles that have been assigned to a group that the user is a member of.

This could lead to the user getting the confirmation prompt (causing confusion) but then ignoring and continuing would actually work.

We should consider getting the effective permissions of the user and checking the individual permission.  I'll look to see if there is an API for this and get back to you.

**Review by wbreza — COMMENTED** [review]
Looking better but still has some issues & limitations.  Apart from changing the verbiage of the error message I don't have any blocking feedback.

File: go.mod [hunk 9]
```diff
@@ -53,7 +53,7 @@ require (
 
 )
 
-require github.com/nathan-fiscaletti/consolesize-go v0.0.0-20220204101620-317176b6684d // indirect
+require github.com/nathan-fiscaletti/consolesize-go v0.0.0-20220204101620-317176b6684d
```
Comment (wbreza) [inline go.mod:56]: Was this change intended?

File: cli/azd/pkg/commands/pipeline/pipeline_manager.go (hunk 8 above)
Comment (ellismg) [reply cli/azd/pkg/commands/pipeline/pipeline_manager.go:459]: > We should consider getting the effective permissions of the user and checking the individual permission. I'll look to see if there is an API for this and get back to you.

Interested in seeing how this works out.  I agree that this is the correct check, but it was non obvious to me how to actually make it.  Interested if you can dig something up here, @wbreza.

The other long term option would be for us to take the error returned by the role assignment operation and examine it and then decide if the error was due to invalid permissions and guide the user on how to get a role assignment setup for them to work.

But I do think we capture about 90% of the value with this change as is, and I wouldn't block landing it here if it takes us a while to figure out how to ask ARM "does this principal have this permission on this resource".

File: cli/azd/pkg/commands/pipeline/pipeline_manager.go [hunk 10]
```diff
@@ -420,3 +446,42 @@ func (manager *PipelineManager) Configure(ctx context.Context) (
...
+	if err != nil {
+		return err
+	}
+	for _, name := range roles {
+		if name == "Owner" || name == "User Access Administrator" {
+			return nil
+		}
+	}
+
+	message := fmt.Sprintf(
```
Comment (ellismg) [inline cli/azd/pkg/commands/pipeline/pipeline_manager.go:469]: I think we can improve the message here a little.  The built in roles most likely exist, it's just that the current user who is running `pipeline config` is not assigned them.

Perhaps something like:

"warning: Your user account does not appear to have either the Owner or User Access Administrator role assigned to it.  This operation will attempt to grant permissions to the newly created service principal and this will  fail unless you have been granted these permissions with a custom role assignment.  Do you want to continue with configuring your pipeline?"

**Review by ellismg — APPROVED** [review]
Some feedback in line with @wbreza's around some of the wording of the errors, but all and all this looks good and should help some folks detect common cases where this operation is going to fail.

Thanks for the improvement here, @hemarina!

File: cli/azd/pkg/commands/pipeline/pipeline_manager.go [hunk 11]
```diff
@@ -420,3 +447,45 @@ func (manager *PipelineManager) Configure(ctx context.Context) (
...
+		Message: message,
+	})
+	if err != nil {
+		return err
+	}
+
+	if !confirm {
+		return fmt.Errorf("missing required user role assignment for authorization: Owner or User Access Administrator. " +
+			"Please contact your Azure subscription administrator to assign required role to your account or" +
+			"login to azd with an account that has required role assignment")
```
Comment (weikanglim) [inline cli/azd/pkg/commands/pipeline/pipeline_manager.go:486]: missing a space between 'or' and 'login'

File: cli/azd/pkg/commands/pipeline/pipeline_manager.go [hunk 12]
```diff
@@ -117,6 +127,23 @@ func (i *PipelineManager) preConfigureCheck(ctx context.Context, infraOptions pr
 		)
 	}
 
+	tenantId, err := i.subResolver.LookupTenant(ctx, i.Environment.GetSubscriptionId())
+	if err != nil {
+		return configurationWasUpdated, fmt.Errorf("getting tenant id for subscription %s. Error: %w",
+			i.Environment.GetSubscriptionId(), err)
```
Comment (weikanglim) [inline cli/azd/pkg/commands/pipeline/pipeline_manager.go:133]: I would simply return `err` here. The underlying error messages already communicate enough details.

File: cli/azd/pkg/commands/pipeline/pipeline_manager.go [hunk 13]
```diff
@@ -420,3 +447,45 @@ func (manager *PipelineManager) Configure(ctx context.Context) (
...
+	confirm, err := console.Confirm(ctx, input.ConsoleOptions{
+		Message: message,
+	})
+	if err != nil {
+		return err
+	}
+
+	if !confirm {
+		return fmt.Errorf("missing required user role assignment for authorization: Owner or User Access Administrator. " +
+			"Please contact your Azure subscription administrator to assign required role to your account or" +
```
Comment (weikanglim) [inline cli/azd/pkg/commands/pipeline/pipeline_manager.go:485]:
```suggestion
			"Please contact your Azure subscription administrator to assign required role(s) to your account or" +
```

File: cli/azd/pkg/commands/pipeline/pipeline_manager.go (hunk 11 above)
Comment (weikanglim) [inline cli/azd/pkg/commands/pipeline/pipeline_manager.go:486]:
```suggestion
			" login to azd with an account that has required role assignment")
```

missing a space between 'or' and 'login'

File: cli/azd/pkg/tools/azcli/ad.go [hunk 14]
```diff
@@ -367,3 +372,68 @@ func (cli *azCli) createRoleAssignmentsClient(
...
+	if err != nil {
+		return nil, err
+	}
+
+	for _, id := range roleDefinitionID {
+		roleDefinitionResult, err := client.GetByID(ctx, id, nil)
+		if err != nil {
+			return nil, fmt.Errorf("failing to get role definition name from role definition id: %w", err)
+		}
+		roleDefinitionName = append(roleDefinitionName, *roleDefinitionResult.Properties.RoleName)
```
Comment (weikanglim) [inline cli/azd/pkg/tools/azcli/ad.go:435]: Can a role definition have an optional name? If so, `RoleName` may be `nil` here and requires handling.

**Review by weikanglim — APPROVED** [review]
